oauth2-server/src/test/java/com/clouway/oauth2/RefreshTokenForClientTest.java

```diff
@@ -28,34 +36,139 @@
+
       oneOf(tokens).refreshToken("::refresh_token::", anyTime);
       will(returnValue(
-              new TokenResponse(true, aNewToken().withValue("::access_token::").expiresAt(anyTime.plusSeconds(600)).build(), "::refresh_token::")
+              new TokenResponse(true,
+                      aNewToken()
+                              .withValue("::access_token::")
+                              .identityId("::identityId::")
+                              .grantType(GrantType.AUTHORIZATION_CODE)
+                              .expiresAt(anyTime.plusSeconds(600))
+                              .build(),
+                      "::refresh_token::")
       ));
+
+      oneOf(identityFinder).findIdentity("::identityId::", GrantType.AUTHORIZATION_CODE, anyTime, Collections.<String, String>emptyMap());
+      will(returnValue(Optional.of(identity)));
+
+      oneOf(request).header("Host");
+      will(returnValue("::host::"));
+
+      oneOf(idTokenFactory).create("::host::", client.id, identity, 600L, anyTime);
+      will(returnValue(Optional.of("::id_token::")));
     }});
 
-    Response response = action.execute(client, new ParamRequest(ImmutableMap.of("refresh_token", "::refresh_token::")), anyTime);
+    Response response = action.execute(client, request, anyTime);
 
     String body = new RsPrint(response).printBody();
 
     assertThat(body, containsString("::access_token::"));
     assertThat(body, containsString("600"));
     assertThat(body, containsString("::refresh_token::"));
+    assertThat(body, containsString("::id_token::"));
+  }
+
+  @Test
+  public void idTokenWasNotGenerated() throws Exception {
+    RefreshTokenActivity action = new RefreshTokenActivity(tokens, idTokenFactory, identityFinder);
+    final Client client = aNewClient().withId("client1").withSecret("secret1").build();
+    final Identity identity = aNewIdentity().withId("::identityId::").build();
+    final DateTime anyTime = new DateTime();
+
+    context.checking(new Expectations() {{
+      oneOf(request).param("refresh_token");
+      will(returnValue("::refresh_token::"));
+
+      oneOf(tokens).refreshToken("::refresh_token::", anyTime);
+      will(returnValue(
+              new TokenResponse(true,
+                      aNewToken()
+                              .withValue("::access_token::")
+                              .identityId("::identityId::")
+                              .grantType(GrantType.AUTHORIZATION_CODE)
+                              .expiresAt(anyTime.plusSeconds(600))
+                              .build(),
+                      "::refresh_token::")
+      ));
+
+      oneOf(identityFinder).findIdentity("::identityId::", GrantType.AUTHORIZATION_CODE, anyTime, Collections.<String, String>emptyMap());
+      will(returnValue(Optional.of(identity)));
+
+      oneOf(request).header("Host");
+      will(returnValue("::host::"));
+
+      oneOf(idTokenFactory).create("::host::", client.id, identity, 600L, anyTime);
```

Comment:
identity is absent why you are trying to generated token for that identity ?

Reply:
I think you are not looking at the right place. identityNotFound test begins at line 138 and at line 160 im returning optional absent and asserting that the body contans "identity was not found"